<center>
    <img src="https://gitlab.com/ibm/skills-network/courses/placeholder101/-/raw/master/labs/module%201/images/IDSNlogo.png" width="300" alt="cognitiveclass.ai logo"  />
</center>


# **Launch Sites Locations Analysis with Folium**


Estimated time needed: **40** minutes


The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.


In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using `matplotlib` and `seaborn` and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using `Folium`.


## Objectives


This lab contains the following tasks:

*   **TASK 1:** Mark all launch sites on a map
*   **TASK 2:** Mark the success/failed launches for each site on the map
*   **TASK 3:** Calculate the distances between a launch site to its proximities

After completed the above tasks, you should be able to find some geographical patterns about launch sites.


Let's first import required Python packages for this lab:


In [1]:
!pip3 install folium
!pip3 install wget

In [2]:
import folium
import wget
import pandas as pd

In [3]:
# Import folium MarkerCluster plugin
from folium.plugins import MarkerCluster
# Import folium MousePosition plugin
from folium.plugins import MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

If you need to refresh your memory about folium, you may download and refer to this previous folium lab:


[Generating Maps with Python](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module\_3/DV0101EN-3-5-1-Generating-Maps-in-Python-py-v2.0.ipynb)


## Task 1: Mark all launch sites on a map


First, let's try to add each site's location on a map using site's latitude and longitude coordinates


The following dataset with the name `spacex_launch_geo.csv` is an augmented dataset with latitude and longitude added for each site.


In [4]:
# Download and read the `spacex_launch_geo.csv`
spacex_csv_file = wget.download('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv')
spacex_df=pd.read_csv(spacex_csv_file)

100% [................................................................................] 7710 / 7710

Now, you can take a look at what are the coordinates for each site.


In [5]:
# Select relevant sub-columns: `Launch Site`, `Lat(Latitude)`, `Long(Longitude)`, `class`
spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index=False).first()
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.


We first need to create a folium `Map` object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.


In [6]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=9) #original zoom: 10

We could use `folium.Circle` to add a highlighted circle area with a text label on a specific coordinate. For example,


In [7]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
    )
)
site_map.add_child(circle)
site_map.add_child(marker)

and you should find a small yellow circle near the city of Houston and you can zoom-in to see a larger circle.


Now, let's add a circle for each launch site in data frame `launch_sites`


*TODO:*  Create and add `folium.Circle` and `folium.Marker` for each launch site on the site map


An example of folium.Circle:


`folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(...))`


An example of folium.Marker:


`folium.map.Marker(coordinate, icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0), html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'label', ))`


In [8]:
coord_site1 = launch_sites_df['Lat'][0],launch_sites_df['Long'][0]
coord_site2 = launch_sites_df['Lat'][1],launch_sites_df['Long'][1]
coord_site3 = launch_sites_df['Lat'][2],launch_sites_df['Long'][2]
coord_site4 = launch_sites_df['Lat'][3],launch_sites_df['Long'][3]
coord_site_list = [coord_site1,coord_site2,coord_site3,coord_site4]
site_names = launch_sites_df['Launch Site']
site_names

0     CCAFS LC-40
1    CCAFS SLC-40
2      KSC LC-39A
3     VAFB SLC-4E
Name: Launch Site, dtype: object

In [9]:
def add_circles(my_map, my_radius):
    for k,coord in enumerate(coord_site_list):
        coordinate = coord
        circle = folium.Circle(coordinate, radius=my_radius, color='#d35400', fill=True)
        circle.add_child(folium.Popup(site_names[k]))
        marker = folium.map.Marker(
            coordinate,
            # Create an icon as a text label
            icon=DivIcon(
                icon_size=(20,20),
                icon_anchor=(0,0),
                html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % site_names[k],
            )
        )
        my_map.add_child(circle)
        my_map.add_child(marker)
    return my_map

In [10]:
enter_view = [coord_site1[0]+5,coord_site1[1]-15]
sites = folium.Map(location=enter_view, zoom_start=4)
add_circles(sites,1000)

The generated map with marked launch sites should look similar to the following:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_markers.png" />
</center>


Now, you can explore the map by zoom-in/out the marked areas
, and try to answer the following questions:

*   Are all launch sites in proximity to the Equator line?
    * ***No. Three of the four sites are near the Tropic of Cancer, though.***
*   Are all launch sites in very close proximity to the coast?
    * ***Yes. The four sites are located in coastal regions. Three by the Atlantic and one by the Pacific.***

Also please try to explain your findings.


# Task 2: Mark the success/failed launches for each site on the map


Next, let's try to enhance the map by adding the launch outcomes for each site, and see which sites have high success rates.
Recall that data frame spacex_df has detailed launch records, and the `class` column indicates if this launch was successful or not


In [11]:
spacex_df.tail(10)

Launch Site        Lat       Long  class
46    KSC LC-39A  28.573255 -80.646895      1
47    KSC LC-39A  28.573255 -80.646895      1
48    KSC LC-39A  28.573255 -80.646895      1
49  CCAFS SLC-40  28.563197 -80.576820      1
50  CCAFS SLC-40  28.563197 -80.576820      1
51  CCAFS SLC-40  28.563197 -80.576820      0
52  CCAFS SLC-40  28.563197 -80.576820      0
53  CCAFS SLC-40  28.563197 -80.576820      0
54  CCAFS SLC-40  28.563197 -80.576820      1
55  CCAFS SLC-40  28.563197 -80.576820      0

Next, let's create markers for all launch records.
If a launch was successful `(class=1)`, then we use a green marker and if a launch was failed, we use a red marker `(class=0)`


Note that a launch only happens in one of the four launch sites, which means many launch records will have the exact same coordinate. Marker clusters can be a good way to simplify a map containing many markers having the same coordinate.


Let's first create a `MarkerCluster` object


In [12]:
marker_cluster = MarkerCluster()

*TODO:* Create a new column in `launch_sites` dataframe called `marker_color` to store the marker colors based on the `class` value


In [13]:
spacex_df['marker_color'] = spacex_df['class'].map({1:'green',0:'red'})
spacex_df.tail()
# Apply a function to check the value of `class` column
# If class=1, marker_color value will be green
# If class=0, marker_color value will be red

Launch Site        Lat      Long  class marker_color
51  CCAFS SLC-40  28.563197 -80.57682      0          red
52  CCAFS SLC-40  28.563197 -80.57682      0          red
53  CCAFS SLC-40  28.563197 -80.57682      0          red
54  CCAFS SLC-40  28.563197 -80.57682      1        green
55  CCAFS SLC-40  28.563197 -80.57682      0          red

In [14]:
# Function to assign color to launch outcome
# def assign_marker_color(launch_outcome):
#     if launch_outcome == 1:
#         return 'green'
#     else:
#         return 'red'
    
# spacex_df['marker_color'] = spacex_df['class'].apply(assign_marker_color)
# spacex_df.tail(10)

*TODO:* For each launch result in `spacex_df` data frame, add a `folium.Marker` to `marker_cluster`


In [15]:
# TODO: Create and add a Marker cluster to the site map
def add_markers(my_map):
    marker_cluster = MarkerCluster()
    my_map.add_child(marker_cluster)
    for index, record in spacex_df.iterrows():
        marker = folium.Marker([record['Lat'], record['Long']],
                               icon=folium.Icon(color='white', icon_color=record['marker_color']))
        marker_cluster.add_child(marker)
    return my_map

In [16]:
enter_view = [coord_site1[0]+5,coord_site1[1]-15]
sites = folium.Map(location=enter_view, zoom_start=4)
add_markers(add_circles(sites,1000))

In [17]:
enter_view = coord_site2
sites = folium.Map(location=enter_view, zoom_start=18)
add_markers(add_circles(sites,50))

Your updated map may look like the following screenshots:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster.png" />
</center>


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster_zoomed.png" />
</center>


From the color-labeled markers in marker clusters, you should be able to easily identify which launch sites have relatively high success rates.


# TASK 3: Calculate the distances between a launch site to its proximities


Next, we need to explore and analyze the proximities of launch sites.


Let's first add a `MousePosition` on the map to get coordinate for a mouse over a point on the map. As such, while you are exploring the map, you can easily find the coordinates of any points of interests (such as railway)


In [18]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
) 
# Use .add_child(mouse_position) at a map to activate mouse position

In [19]:
enter_view = [coord_site1[0]+5,coord_site1[1]-15]
sites = folium.Map(location=enter_view, zoom_start=4)
add_markers(add_circles(sites,1000)).add_child(mouse_position)

Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


In [20]:
enter_view = coord_site4
sites = folium.Map(location=enter_view, zoom_start=14)
add_markers(add_circles(sites,50)).add_child(mouse_position)

You can calculate the distance between two points on the map based on their `Lat` and `Long` values using the following method:


In [21]:
from math import sin, cos, sqrt, atan2, radians, degrees

def calculate_distance(*args): 
    '''
    Arguments: lat1, lon1, lat2, lon2.
    '''
    b = args
    b = [tuple(b[i]) if type(b[i]) in [tuple,list] else (b[i],) for i in range(0,len(b))]
    b = sum(b,())
    if len(b) == 4:
        lat1 = radians(b[0])
        lon1 = radians(b[1])
        lat2 = radians(b[2])
        lon2 = radians(b[3])
    else:
        return print(f'Wrong number of arguments: need 4, got {len(b)}.')
    dlon = lon2 - lon1
    dlat = lat2 - lat1
    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))
    R = 6373.0 # approximate radius of earth in km
    distance = R * c
    return distance

*TODO:* Mark down a point on the closest coastline using MousePosition and calculate the distance between the coastline point and the launch site.


In [23]:
# Distance to the coastline from launch site VAFB SLC-4E
# Lat: 34.63568 Long: -120.62542
coord_coast = 34.63568, -120.62542
distance_coast = calculate_distance(coord_site4, coord_coast) 
distance_coast

1.3798034019347332

*TODO:* After obtained its coordinate, create a `folium.Marker` to show the distance


In [24]:
# Create and add a folium.Marker on your selected closest coastline point on the map
# Display the distance between coastline point and launch site using the icon property 
# for example
distance_marker_coast = folium.Marker(
    coord_coast,
    icon=DivIcon(
        icon_size=(20,20), icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_coast),
    )
)
# Use .add_child(distance_marker_coast) at a map to add distance marker

*TODO:* Draw a `PolyLine` between a launch site to the selected coastline point


In [25]:
# Create a `folium.PolyLine` object using the coastline coordinates and launch site coordinate
line_coast = folium.PolyLine(locations=[coord_site4, coord_coast], weight=1)
# Use .add_child(line_coast) at a map to add a line

In [26]:
enter_view = coord_site4
sites = folium.Map(location=enter_view, zoom_start=15)
sites = add_markers(add_circles(sites,200))
sites.add_child(mouse_position)
sites.add_child(distance_marker_coast)
sites.add_child(line_coast)

Your updated map with distance line should look like the following screenshot:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_distance.png" />
</center>


*TODO:* Similarly, you can draw a line betwee a launch site to its closest city, railway, highway, etc. You need to use `MousePosition` to find the their coordinates on the map first


A railway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/railway.png" />
</center>


A highway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/highway.png" />
</center>


A city map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/city.png" />
</center>


In [27]:
# Create a marker with distance to a closest city, railway, highway, etc.
# Draw a line between the marker to the launch site

In [28]:
# Railroad near launch site CCAFS SLC-40 (row [1], site2):
# Lat: 28.57209 Long: -80.58529
coord_rail = 28.57209, -80.58529
distance_rail = calculate_distance(coord_site2,coord_rail)
distance_marker_rail = folium.Marker(
    coord_rail,
    icon=DivIcon(
        icon_size=(20,20), icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_rail),
    )
)
line_rail = folium.PolyLine(locations=[coord_site2, coord_rail], weight=1)

In [29]:
enter_view = coord_site2
sites = folium.Map(location=enter_view, zoom_start=14)
sites = add_markers(add_circles(sites,200))
sites.add_child(mouse_position)
sites.add_child(distance_marker_rail)
sites.add_child(line_rail)

In [30]:
# City near launch site CCAFS LC-40 (row [0], site1):
# Lat: 28.40208 Long: -80.60236
coord_city = 28.40208, -80.60236
distance_city = calculate_distance(coord_site1, coord_city)
distance_marker_city = folium.Marker(
    coord_city,
    icon=DivIcon(
        icon_size=(20,20), icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_city),
    )
)
line_city = folium.PolyLine(locations=[coord_site1, coord_city], weight=1)

In [31]:
enter_view = coord_site1
adjust_lat = -0.1
enter_view = enter_view[0]+adjust_lat, enter_view[1]
sites = folium.Map(location=enter_view, zoom_start=11)
sites = add_markers(add_circles(sites,200))
sites.add_child(mouse_position)
sites.add_child(distance_marker_city)
sites.add_child(line_city)

In [32]:
# Highway near launch site KSC LC-39A (row [2], site3):
# Lat: 28.57352 Long: -80.65546
coord_high = 28.57352, -80.65546
distance_high = calculate_distance(coord_site3,coord_high)
distance_marker_high = folium.Marker(
    coord_high,
    icon=DivIcon(
        icon_size=(20,20), icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_high),
    )
)
line_high = folium.PolyLine(locations=[coord_site3, coord_high], weight=1)

In [33]:
enter_view = coord_site3
sites = folium.Map(location=enter_view, zoom_start=15)
sites = add_markers(add_circles(sites,200))
sites.add_child(mouse_position)
sites.add_child(distance_marker_high)
sites.add_child(line_high)

After you plot distance lines to the proximities, you can answer the following questions easily:

*   Are launch sites in close proximity to railways?
    * ***Yes. Launch site CCAFS SLC-40 is only 1.3 km away from the nearest railroad, for example.***
*   Are launch sites in close proximity to highways?
    * ***Yes. Launch site KSC LC-39A is only about 0.8 km away from the nearest highway, for example.***
*   Are launch sites in close proximity to coastlines?
    * ***Yes. Launch site VAFB SLC-4E is only 1.4 km away from coastline, for example.***
*   Do launch sites keep certain distance away from cities?
    * ***Yes. Launch site CCAFS LC-40 is almost 18 km away from Cape Canaveral, for example.*** 

Also please try to explain your findings.


# Next Steps:

Now you have discovered many interesting insights related to the launch sites' location using folium, in a very interactive way. Next, you will need to build a dashboard using Ploty Dash on detailed launch records.


## Authors


[Yan Luo](https://www.linkedin.com/in/yan-luo-96288783/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2021-01-01)


### Other Contributors


Joseph Santarcangelo


## Change Log


| Date (YYYY-MM-DD) | Version | Changed By | Change Description          |
| ----------------- | ------- | ---------- | --------------------------- |
| 2021-05-26        | 1.0     | Yan        | Created the initial version |


Copyright © 2021 IBM Corporation. All rights reserved.
